[docs] Rework the pickers Getting Started page (#7140)
mui/mui-x: docs/data/date-pickers/getting-started/getting-started.md

The simplest way to use `LocalizationProvider` is to wrap your entire application in it.
You will then be able to use the Date and Time Pickers everywhere.

{{"component": "modules/components/PickersRenderingInstructions.js"}}

:::success

- All the adapters are exported by both `@mui/x-date-pickers` and `@mui/x-date-pickers-pro`:

  ```tsx
  import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns';
  import { AdapterDateFns } from '@mui/x-date-pickers-pro/`AdapterDayjs`';
  ```

- `LocalizationProvider` is exported by both `@mui/x-date-pickers` and `@mui/x-date-pickers-pro`:

  ```tsx
  import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
  import { LocalizationProvider } from '@mui/x-date-pickers-pro/LocalizationProvider';
  ```

:::

:::info
If you need to use the Date and Time Pickers with a custom locale, have a look at the [Localized dates](/x/react-date-pickers/adapters-locale/) page.
:::

## Render your first component

To make sure that everything is set up correctly, try rendering a simple `DatePicker` component:

> {{"demo": "FirstComponent.js"}}

Author: > and wrap with `LocalizationProvider` on the application level?

How would you do it ?
For now, the doc does not set a `LocalizationProvider` outside of the demo.
We could add one to enable demos without `LocalizationProvider`, if we consider that the doc is clear enough about how it should be used (it won't be possible to copy paste a demo, just like with the Joy demos which require the CSS Variables context).

```sh
// with npm
npm install @mui/material @mui/styled-engine-sc styled-components

// with yarn
yarn add @mui/material @mui/styled-engine-sc styled-components
```

## Code setup

Before trying to render any component, you have to make sure that there is a `LocalizationProvider` upper in the React tree.
This component receives your chosen date library's adapter and makes it accessible to all the Date and Time Pickers component using React context.

Each demonstration in the documentation has its own `LocalizationProvider` wrapper.
This is **not** a pattern to reproduce.
The reason is to keep examples atomic and functional—especially when running in a CodeSandbox.

The general recommendation is to declare the `LocalizationProvider` once, wrapping your entire application.
Then, you don't need to repeat the boilerplate code for every Date and Time Picker in your application.

{{"component": "modules/components/PickersRenderingInstructions.js"}}

:::success

- All the adapters are exported by both `@mui/x-date-pickers` and `@mui/x-date-pickers-pro`:

  ```tsx
  import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';

>   import { AdapterDayjs } from '@mui/x-date-pickers-pro/`AdapterDayjs`';

Reviewer: ```suggestion
  import { AdapterDayjs } from '@mui/x-date-pickers-pro/AdapterDayjs';
```

waiAria: https://www.w3.org/WAI/ARIA/apg/example-index/dialog-modal/datepicker-dialog.html
---

# Date and Time Pickers

<p class="description">The Date and Time pickers let the user select date and time values.</p>

{{"component": "modules/components/ComponentLinkHeader.js"}}

## Installation

> To make the Date and Time Pickers, work, you need to install the following packages:

Reviewer 2: ```suggestion
To make the Date and Time Pickers work, you need to install the following packages:
```

1. **The component library** (`@mui/x-date-pickers` or `@mui/x-date-pickers-pro`) to render the component.
2. **The date library** ([Day.js](https://day.js.org/), [date-fns](https://date-fns.org/), ...) to manipulate the date.

> {{"component": "modules/components/PickersInstallationInstructions.js"}}

Author: My approach here was to consider that the vast majority of people would understand the form without reading all those long sections.
They see "Date library", they click, they see there favorite date library, they install.

With that approach, the interactive installation is the main element and the sections and a notice guide coming after.

Other wise, the section ends up super low and is easy to miss.

Before trying to render any component, you have to make sure that there is a `LocalizationProvider` upper in the React tree.
This component receives your chosen date library's adapter and makes it accessible to all the Date and Time Pickers component using React context.

> Most of the demos in the documentation use the `LocalizationProvider` wrapping only the components being demoed. That's because we aim to keep our examples atomic and functional with copy and paste.

Reviewer 2: ```suggestion
Each demonstration in the documentation has its own `LocalizationProvider` wrapper.
This is **not** a pattern to reproduce.
The reason is to keep examples atomic and functional.
Especially when running in a CodeSandbox.
```

  ```

- `LocalizationProvider` is exported by both `@mui/x-date-pickers` and `@mui/x-date-pickers-pro`:

  ```tsx
  import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
  import { LocalizationProvider } from '@mui/x-date-pickers';
  import { LocalizationProvider } from '@mui/x-date-pickers-pro/LocalizationProvider';
  import { LocalizationProvider } from '@mui/x-date-pickers-pro/';
  ```

:::

:::info
If you need to use the Date and Time Pickers with a custom locale, have a look at the [Localized dates](/x/react-date-pickers/adapters-locale/) page.
:::

## Commonly used components

Most applications only need a few of the components exposed by the Date and Time packages:

{{"demo": "CommonlyUsedComponents.js"}}

:::success
All the components exported by `@mui/x-date-pickers` are also exported by `@mui/x-date-pickers-pro` but not with nested imports.

For example, to use the `DatePicker` component, the three following imports are valid:

```tsx
import { DatePicker } from '@mui/x-date-pickers/DatePicker';
import { DatePicker } from '@mui/x-date-pickers';
import { DatePicker } from '@mui/x-date-pickers-pro';
```

:::

> ## Basic use cases

Reviewer 2: I agree that "form props" can be discovered by reading deeper docs, or simply by typing `disabled` in the components like for all the MUI components 😉

The results may vary in your application depending on the version of each library, the locale, and the bundling strategy used.
:::

### Other peer dependencies

The Date and Time Pickers package has a peer dependency on `@mui/material`.

> If you are not already using it in your project, you can install it with:

Author: I think `you can` means `your are able to` here not `you have the possibility to`.
It's @samuelsycamore wording on the grid equivalent page, if we change on this one we should change there to.

```sh
// with npm
npm install @mui/material @emotion/react @emotion/styled

// with yarn
yarn add @mui/material @emotion/react @emotion/styled
```

> 
> <!-- #react-peer-version -->

Reviewer 2: Does it have an interest for maintenance of the docs?
```suggestion
```

Author: No idea, I copy pasted the grid doc :laughing:

Please note that [react](https://www.npmjs.com/package/react) >= 17.0.2 and [react-dom](https://www.npmjs.com/package/react-dom) >= 17.0.2 are peer dependencies.

MUI is using [emotion](https://emotion.sh/docs/introduction) as a styling engine by default. If you want to use [`styled-components`](https://styled-components.com/) instead, run:

```sh
// with npm
npm install @mui/material @mui/styled-engine-sc styled-components

// with yarn
yarn add @mui/material @mui/styled-engine-sc styled-components
```

## Code setup

Before trying to render any component, you have to make sure that there is a `LocalizationProvider` upper in the React tree.
This component receives your chosen date library's adapter and make it accessible to all the Date and Time Pickers component with a React context.

> The most simple way to use `LocalizationProvider` is to wrap your entire application it.

Reviewer 2: ```suggestion
The simplest way to use `LocalizationProvider` is to wrap your entire application in it.
```

You will then be able to use the Date and Time Pickers everywhere.

{{"component": "modules/components/PickersRenderingInstructions.js"}}

:::success

- All the adapters are exported by both `@mui/x-date-pickers` and `@mui/x-date-pickers-pro`:

  ```tsx
  import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns';
  import { AdapterDateFns } from '@mui/x-date-pickers-pro/`AdapterDayjs`';
  ```

- `LocalizationProvider` is exported by both `@mui/x-date-pickers` and `@mui/x-date-pickers-pro`:

  ```tsx
  import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
  import { LocalizationProvider } from '@mui/x-date-pickers';
  import { LocalizationProvider } from '@mui/x-date-pickers-pro/LocalizationProvider';
  import { LocalizationProvider } from '@mui/x-date-pickers-pro/';
  ```

:::

:::info
If you need to use the Date and Time Pickers with a custom locale, have a look at the [Localized dates](/x/react-date-pickers/adapters-locale/) page.
:::

## Commonly used components

Most applications only need a few of the components exposed by the Date and Time packages:

> {{"demo": "CommonlyUsedComponents.js"}}

Reviewer 2: Amazing demo for introducing and playing with components 🎉

But it's the first time reader encounter a demo, and the code is complicated to read.

Author: Yeah, this demo is interesting to help people find quickly their component, not to help them copy paste code...
We could split into 4 demos (this page has a ton of very basic demo anyway)

But my idea was that it was the "Controlled value" demo which should be the entry point for people to copy / paste, since it's closer from a real use case scenario.
I do agree that it's not very accessible right now though.

:::success
All the components exported by `@mui/x-date-pickers` are also exported by `@mui/x-date-pickers-pro` but not with nested imports.

For example, to use the `DatePicker` component, the three following imports are valid:

```tsx
import { DatePicker } from '@mui/x-date-pickers/DatePicker';
import { DatePicker } from '@mui/x-date-pickers';
import { DatePicker } from '@mui/x-date-pickers-pro';
```

:::

## Basic use cases

### Controlled value

All the components can be controlled using the `value` / `onChange` props:

{{"demo": "ControlledComponent.js"}}

### Custom format

All the components with an HTML input can receive a custom format using the [`format` prop](/x/react-date-pickers/adapters-locale/#custom-field-format):

{{"demo": "CustomFormat.js"}}

### Form props

All the components component can be disabled or read-only:

{{"demo": "FormProps.js"}}

### Validation

If you need to add validation to your component, you can check the [Validation page](/x/react-date-pickers/validation/):

{{"demo": "ValidatedComponent.js"}}

## Other components

> ### Keyboard or mouse editing

Reviewer 2: "User interaction"?

> ### Keyboard or mouse editing

Author: I went for `Choose interaction style`, let's see what Sam proposes :+1:

The first thing you want to decide, is how the user should be able to select their value:

- For keyboard and mouse editing, use the _Picker_ components:

{{"demo": "BasicDatePicker.js", "hideToolbar": true, "bg": "inline"}}

- For keyboard-only editing, use the _Field_ components:

{{"demo": "BasicDateField.js", "hideToolbar": true, "bg": "inline"}}

> - For mouse-only editing, use the _Calendar / Clock_ components:

Author: I made a proposal on the new `introduction.md` page
Still note convinced but that's better

{{"demo": "BasicDateCalendar.js", "hideToolbar": true, "bg": "inline"}}

:::success
Each _Picker_ is the combination of one _Field_ and one or several _Calendar / Clock_ components.
For example, the `DatePicker` is the combination of the `DateField` and the `DateCalendar`.

The _Calendar / Clock_ components are rendered inside a _Popover_ on desktop and inside a _Modal_ on mobile.
:::

### Date or time editing ?

The Date and Time Pickers are divided into six families of components.

> The demo below shows each one of them using their field component:

Author: I'm using the field to use the same type of component on each family but we can replace with the pickers when they are all available.

I don't think it's a problem because it's an advanced section, nobody should reach this section when just looking for a `DatePicker`.

Reviewer: Fair point. I don't see problems with it. 🤔

{{"demo": "ComponentFamilies.js"}}

> ### Responsiveness

Reviewer: Hm...
This would create an inconsistency between our new approach of naming such section in individual docs as `Available components`.

> ### Responsiveness

Author: True, but if we use `Available components` it's super unclear in the context of this page.

On the `date-picker.md`, `Available components` list all the components in the scope of the page.
That's not the case here, we are just talking about the component dedicated to the various type of device.

If at one point we migration to the modal vs popover naming, then we could have a name describing this behavior.

> ### Responsiveness

Reviewer 2: Yes, you could skip them. Their main use case is if you need to create advanced components where you override the Modal/Popper. So it can be deeper in the docs

> ### Responsiveness

Author: If you remove entirely this section, would you have no place to describe our responsive behaviors outside of the component pages ? 

That can make sense
Or we could, in the future, have a page describing more in depth how we handle responsiveness (a boosted version of the `Available components` sections but with examples using `desktopMediaQuery`)

> Each _Picker_ is available in three responsive variants:

Reviewer: Technically, they all have only one _responsive_ variant. 🤔 
They can also be used as `desktop` or `mobile` variants as well as `static` variant.
Maybe @samuelsycamore or @joserodolfofreitas would have better ideas on how to "translate" this behaviour/setup of our components into words. 🙈 
For reference: https://next.mui.com/x/react-date-pickers/date-picker/#available-components

- The desktop component (e.g. `DesktopNextDatePicker`) which works best for mouse devices and large screens.
  It renders the views inside a popover and allows editing values directly inside the field.

- The mobile component (e.g. `MobileNextDatePicker`) which works best for touch devices and small screens.
  It renders the view inside a modal and does not allow editing values directly inside the field.

- The responsive component (e.g. `NextDatePicker`) which renders the desktop component or the mobile one depending on the device it runs on.

{{"demo": "ResponsivePickers.js"}}

### Find your component

Use the form below to find component you need:

> {{"demo": "ComponentExplorerNoSnap.js", "hideToolbar": true}}

Reviewer: This is a really nice demo!!! 🚀 ❤️

> {{"demo": "ComponentExplorerNoSnap.js", "hideToolbar": true}}

Reviewer 2: 😍🤩

Importing it from `@mui/x-date-pickers-pro` is enough.
:::

> ## Native pickers

Reviewer: Agreed, IMHO this section made little to no sense in this location...
Maybe it would have usefulness at the beginning of the page, but that is also debatable.
If people are in search of a date library, they probably already know that native solutions lack what they need. 🤷 🙈 💡
